### ISSUE TITLE
ADPCM is not correctly recognized in .act files

### ISSUE THREAD

**Issue by Sami32** [MediaArea/MediaInfo#117]
ADPCM is recognized as LPCM in a .act file (produced by MP3/MP4 recorders)
https://en.wikipedia.org/wiki/ACT_(audio_format)

Here a file sample:
[REC001.ACT.txt](https://github.com/MediaArea/MediaInfo/files/721342/REC001.ACT.txt)

**Comment by JeromeMartinez** [comment]
Hum... I don't like that!
the WAV header says it is PCM 8 kHz:

```
000C  Stream format - Audio (24 bytes)
000C   Header (8 bytes)
000C    Name:                                fmt 
0010    Size:                                16 (0x00000010)
0014   FormatTag:                            1 (0x0001)
0016   Channels:                             1 (0x0001)
0018   SamplesPerSec:                        8000 (0x00001F40)
001C   AvgBytesPerSec:                       16000 (0x00003E80)
0020   BlockAlign:                           2 (0x0002)
0022   BitsPerSample:                        16 (0x0010)
```

FormatTag is 1, AvgBytesPerSec is SamplesPerSec*BitsPerSample/8 as expected for PCM, looks like there is no way to detect ACT without probing the content of the file, but I find no "spec" about ACT.
Do you have something more detailed about ACT format?

Please provide a couple of additional files, maybe there is a "magic number" in the begin of the audio content saying it is ACT.
I could test the file extension but I don't like to rely on file extension, as anyone can change it and there is not always a file name.

**Comment by Sami32** [comment]
Sure, here a couple of samples:
[adpcm.zip](https://github.com/MediaArea/MediaInfo/files/721358/adpcm.zip)
I only saw 8000Hz, 1 channels, 128kb/s constant between them.
> Primary association: ADPCM Compressed File Recorded by Some MP3 Players/Voice Recorders
> Identifying characters Hex: 00 00 FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF , ASCII: 

No sadly, i wasn't able to find any specifications, or maybe in chineese ;) just kidding.

**Comment by Sami32** [comment]
I'm not sure that is relevant, but it's look like they all begin with `52 49 46 46 44 de`   string.

**Comment by JeromeMartinez** [comment]
thanks.
I saw the filext description but this string is actually not in the file.
I also read that open source SoX can decode it, so I guess I could find the corresponding code but it is more time to spend on it.

Looks like there is a kind of specific bytes always 0 up to offset 512, and 4 bytes only with changing content at offset 256, maybe it is enough for detection.

> but it's look like they all begin with 52 49 46 46 44 de string.

first 44 bytes are the classic WAVE file format header ;-).

**Comment by Sami32** [comment]
OK, so i deleted my previous post.
You much competent than me ;)

Thank you for your attention and help :smile:

**Comment by Sami32** [comment]
You're right, the byte at offset 256 is always 84 and the 3 others after changes followed by zero as you already explained.

**Comment by JeromeMartinez** [comment]
For info [FFmpeg probes it in a similar way](https://github.com/FFmpeg/FFmpeg/blob/6f69f7a8bf6a0d013985578df2ef42ee6b1c7994/libavformat/act.c#L40-L58).

**Comment by Sami32** [comment]
That's a smart find :D

Thank you for sharing :+1:

**Comment by Sami32** [comment]
@JeromeMartinez I just discovered `IsTruncated: Yes` field in the recent MediaInfo version, that is displayed only for this chineese proprietary WAV ADPCM format.

Is that intentional ? for this only purpose ?
As i would like use this field for parsing such files, i prefer know excatly what is the purpose of this field.
Thank you.

**Comment by JeromeMartinez** [comment]
It is because the indicate file size is bigger than the real size:

```
0000` Wave (44 bytes)
000C  Stream format - Audio (24 bytes)
000C   Header (8 bytes)
000C    Name:                                fmt 
0010    Size:                                16 (0x00000010)
0014   FormatTag:                            1 (0x0001)
0016   Channels:                             1 (0x0001)
0018   SamplesPerSec:                        8000 (0x00001F40)
001C   AvgBytesPerSec:                       16000 (0x00003E80)
0020   BlockAlign:                           2 (0x0002)
0022   BitsPerSample:                        16 (0x0010)
0024  Raw datas (8 bytes)
0024   Header (8 bytes)
0024    Name:                                data
0028    Size:                                122400 (0x0001DE20)
```

See last size.
I don't know if it is in all files and just that such encoders are crappy and just don't care of specs, or if it is only for some files.
If it is "buggy" everywhere, I guess I should remove this line for ACT.

**Comment by Sami32** [comment]
Please don't :pray:  as i plan to use it on Wave format, to force our parser to considerate them as ADPCM.

Not perfect, but better than rely only on 8000Hz, 1 channels, 128k/s and 16 bit conditions for now i guess.

**Comment by Nadahar** [comment]
Wouldn't it be better to "detect them properly" than relying on that they all have a buggy size indicator? If there are no way to "detect them properly" we might consider just dropping trying to support them.

**Comment by JeromeMartinez** [comment]
Hum. Not the right method: ADPCM should be tested differently, with the "hard coded bytes" probing.
Do not rely at `IsTruncated: Yes` as you can see that in a lot of other files (really truncated).
This would be only a temporary workaround the time I implement the detection of such ACT stuff.

**Comment by Sami32** [comment]
@JeromeMartinez Yes, i was seing this only as a workaround.

Thank you for your insight and understanding :smiley:
